cmds/which/which_linux.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2016-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"os"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	// Effective UID and GID. Technically, Linux checks the fsuid and
+	// fsgid. But those are in 99.999% of cases the same as EUID and EGID.
+	// Nobody uses that moldy stuff anymore.
+	eUID = uint32(os.Geteuid())
+	eGID = uint32(os.Getegid())
+)
+
+func stat(path string) (unix.Stat_t, error) {
+	var s unix.Stat_t
+	if err := unix.Stat(path, &s); err != nil {
+		return unix.Stat_t{}, err
+	}
+	return s, nil
+}
+
+// Execute permission bits.
+const (
+	otherExec = 1
+	groupExec = 1 << 3
+	userExec  = 1 << 6
+)
+
+func canExecute(path string) bool {
```

Comment:
This should just use faccess at, e.g.
faccessat(AT_FDCWD, "/bin/date", X_OK)  = 0

Reply:
faccessat does not to what you want it to do. It won't tell if you can actually exec the file. For one, faccessat uses the real UID/GID, while execve will use the effective UID/GID. the access family of syscalls is mostly useful for setuid programs.